temporaltest/tls.go

```diff
@@ -0,0 +1,165 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the MIT License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
+
+package temporaltest
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"time"
+)
+
+type certData struct {
+	cert  *x509.Certificate
+	bytes []byte
+	key   *rsa.PrivateKey
+}
+
+type Certificate struct {
+	CertPem []byte
+	KeyPem  []byte
+}
+
+type Bundle struct {
+	Ca     *Certificate
+	Client *Certificate
+	Server *Certificate
+}
+
+func GenerateCertificates() (*Bundle, error) {
+	ca, err := generateCa()
+	if err != nil {
+		return nil, err
+	}
+
+	server, err := generateServerCertificate(ca)
+	if err != nil {
+		return nil, err
+	}
+	client, err := generateClientCertificate(ca)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &Bundle{
+		Ca:     convertCertData(ca),
+		Client: convertCertData(client),
+		Server: convertCertData(server),
+	}
+
+	return b, nil
+}
+
+func convertCertData(d *certData) *Certificate {
+	certPem := new(bytes.Buffer)
+	pem.Encode(certPem, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: d.bytes,
+	})
+
+	priv := new(bytes.Buffer)
+	pem.Encode(priv, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(d.key),
+	})
+
+	return &Certificate{
+		CertPem: certPem.Bytes(),
+		KeyPem:  priv.Bytes(),
+	}
+}
+
+func generateCa() (*certData, error) {
+	ca := &x509.Certificate{
+		SerialNumber: big.NewInt(2022),
+		Subject: pkix.Name{
+			Organization:  []string{"Company, INC."},
+			Country:       []string{"US"},
+			Province:      []string{""},
+			Locality:      []string{"San Francisco"},
+			StreetAddress: []string{"Golden Gate Bridge"},
+			PostalCode:    []string{"94016"},
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	private, err := rsa.GenerateKey(rand.Reader, 4096)
```

Comment:
For testing purposes, I would recommend an ECDSA key which is _way_ faster to generate. If you want to see some code I use in the Python SDK for generating such certs, see https://github.com/temporalio/sdk-python/blob/962a5e8bebdf1683f171afac36b7910549301189/tests/helpers/golangserver/certs/gen.go

Reply:
Do you think it makes sense to do both ECDSA and RSA keys?

Reply:
For a testing package that creates both sides, not really. In fact, generating certificates is really just helper code anyone can write in whatever way they prefer, so not sure you even need to export this utility. Just generate your certificates yourself in your test file and let users of this library generate theirs how they want.